File: src/bumperbot_py_examples/bumperbot_py_examples/simple_subscriber.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from std_msgs.msg import String

# create a subscriber class
class SimpleSubscriber(Node): # inherit from Node
    
    def __init__(self): # constructor
        
        # call constructor of the base class (takes String with name of node)
        super().__init__('simple_subscriber') 
        
        # create a subscriber for String messages in channel chatter, 
        # max size 10
        self.sub_ = self.create_subscription(
            String, #message type
            'chatter', # topic name
            self.msgCallback, # callback function
            10 # size of the subscription queue
           ) 

    def msgCallback(self, msg): #define callback as a member of the class
        self.get_logger().info('I heard: %s' % msg.data) # log message
        

def main():

    # 1) Initialization: calling `init` for a particular Context. 
    # Must be done before any ROS nodes can be created.
    rclpy.init()
    

    # 2) Create one or more ROS nodes: by calling `create_node` or instantiating
    # a Node. Nodes may implement common ROS entities such as publishers, 
    # subscriptions, services and actions.
    simple_subscriber = SimpleSubscriber() # create a SimpleSubscriber object

    # 3) Process node callbacks: by calling `spin`, `spin_once`, or 
    # `spin_until_future_complete`.
    # After a node is created, one can process work items (e.g. subscription 
    # callbacks) that are waiting to be executed
    rclpy.spin(simple_subscriber) # keep node running continuously to process callbacks

    # 4) Shutdown
    # When finished with a previously initialized .Context (i.e. done using
    # all ROS nodes associated with the context), the shutdown function should 
    # be called. This will invalidate all entities derived from the context.
    simple_subscriber.destroy_node() # destroy node
    rclpy.shutdown() # shutdown ROS2

if __name__ == '__main__':
    main()

File: src/bumperbot_cpp_examples/src/simple_subscriber.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>

using std::placeholders::_1; // to be able to use _1 as a placeholder

// Define the SimpleSubscriber class as a subclass of rclcpp::Node
class SimpleSubscriber : public rclcpp::Node
{
public:
    // constructor, initialize Node with node name "simple_subscriber"
    SimpleSubscriber() : Node("simple_subscriber")
    {
        // Create a subscriber on the "chatter" topic
        sub_ = this->create_subscription<std_msgs::msg::String>(
            "chatter",                                           // topic name
            10,                                                 // size of buffer queue
            std::bind(&SimpleSubscriber::msgCallback, this, _1) // callback function
        );

        RCLCPP_INFO(get_logger(), "Listening in topic /chatter");
    }

private:
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub_; // shared pointer to the publisher with a String interface
    void msgCallback(const std_msgs::msg::String::SharedPtr msg) const     // callback function to be called at 1s intervals
    {
        RCLCPP_INFO(this->get_logger(), "I heard: '%s'", msg->data.c_str());
    }
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv); // initialize ROS2

    auto node = std::make_shared<SimpleSubscriber>(); // create a shared pointer to the SimpleSubscriber class

    rclcpp::spin(node); // run the node

    rclcpp::shutdown(); // shutdown ROS2

    return 0;
}